tools/clang/lib/Sema/SemaHLSLDiagnoseTU.cpp

```diff
@@ -325,6 +325,96 @@ class HLSLCallDiagnoseVisitor
     return true;
   }
 
+  bool VisitDeclRefExpr(DeclRefExpr *ET) {
+    // Search for enum types that should only exist after
+    // specific shader models.
+    IdentifierInfo *II = ET->getDecl()->getIdentifier();
+
+    if (!SM->IsSMAtLeast(6, 9)) {
+      if (II && II->getName().startswith("RAYQUERY_FLAG")) {
+        sema->Diag(ET->getLocation(), diag::warn_hlsl_enum_type_not_allowed)
+            << II->getName() << SM->GetName() << "6.9";
+      }
+    }
+
+    return true;
+  }
+
+  bool VisitCXXMemberCallExpr(CXXMemberCallExpr *callExpr) {
+    // Diagnose TraceRayInline member call
+    const MemberExpr *ME = dyn_cast<MemberExpr>(callExpr->getCallee());
+
+    if (ME) {
+      const Decl *MD = ME->getMemberDecl();
+
+      if (const CXXMethodDecl *methodDecl = dyn_cast<CXXMethodDecl>(MD)) {
+        if (methodDecl->getNameAsString() == "TraceRayInline") {
+          // validate that if the ray flag parameter is
+          // RAY_FLAG_FORCE_OMM_2_STATE, the ray query object
+          // has the RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS
+          // flag set, otherwise emit a diagnostic
```

Comment:
Is this the "normal" way of doing this?  I'm a bit surprised to see string comparisons here, but maybe that's just how things are done?

Reply:
I've found a way to compare opcodes instead of string comparing.